File: source/algorithm/NQ071/NQ071_improved_improved.cpp
#include <algorithm>
#include <iostream>
using namespace std;

const int MAX_N = 100005;
int N, C, stalls[MAX_N];

bool canPlace(int distance) {
    int count = 1, last = stalls[0];
    for (int i = 1; i < N; i++) {
        if (stalls[i] - last >= distance) {
            count++;
            last = stalls[i];
            if (count >= C)
                return true;
        }
    }
    return false;
}

int main() {
    scanf("%d%d", &N, &C);
    for (int i = 0; i < N; i++) {
        scanf("%d", &stalls[i]);
    }
    sort(stalls, stalls + N);

    int left = 1, right = stalls[N - 1] - stalls[0];
    while (left < right) {
        int mid = left + (right - left + 1) / 2;
        if (canPlace(mid)) {
            left = mid;
        } else {
            right = mid - 1;
        }
    }
    printf("%d\n", left);

    return 0;
}

File: 兴趣班入门百练讲义/NQ100/NQ071/NQ071_improved_improved.py
def can_place(stalls, distance, C):
    count = 1
    last = stalls[0]
    for i in range(1, len(stalls)):
        if stalls[i] - last >= distance:
            count += 1
            last = stalls[i]
            if count >= C:
                return True
    return False

def main():
    import sys
    input = sys.stdin.read().split()
    idx = 0
    N = int(input[idx])
    idx += 1
    C = int(input[idx])
    idx += 1
    stalls = []
    for _ in range(N):
        stalls.append(int(input[idx]))
        idx += 1
    stalls.sort()
    
    left = 1
    right = stalls[-1] - stalls[0]
    while left < right:
        mid = left + (right - left + 1) // 2
        if can_place(stalls, mid, C):
            left = mid
        else:
            right = mid - 1
    print(left)

if __name__ == "__main__":
    main()
